### PR TITLE
fix(coderd): ensure correct RBAC when enqueueing notifications

### PR DESCRIPTION
**Pull request by johnstcn**
While investigating https://github.com/coder/coder/issues/15477 I noticed the following on my test workspace:

```
2024-11-11 16:10:01.076 [warn]  autobuild: failed to notify of workspace marked as dormant  workspace_id=3a5b3ab6-c0f6-496f-8efe-25e02a2a9885  workspace_name=apricot-starfish-90  workspace_id=3a5b3ab6-c0f6-496f-8efe-25e02a2a9885 ...
    error= new message metadata:
               github.com/coder/coder/v2/coderd/notifications.(*StoreEnqueuer).EnqueueWithData
                   /home/runner/work/coder/coder/coderd/notifications/enqueuer.go:69
             - unauthorized: rbac: forbidden
```

I expanded the search and found that this was a _fundamental_ problem with our test notifier implementation -- namely, that it made absolutely no RBAC assertions.

[Logs (loki)](https://grafana.dev.coder.com/explore?schemaVersion=1&panes=%7B%22giu%22%3A%7B%22datasource%22%3A%22loki%22%2C%22queries%22%3A%5B%7B%22refId%22%3A%22A%22%2C%22expr%22%3A%22sum%28count_over_time%28%5Cn++++%7Bnamespace%3D%5C%22coder%5C%22%2C+container%3D%5C%22coder%5C%22%2C+pod%21%7E%5C%22coder-provisioner.*%5C%22%7D+%7C%3D+%60unauthorized%3A+rbac%3A+forbidden%60+%7C%3D+%60%2Fhome%2Frunner%2Fwork%2Fcoder%2Fcoder%2Fcoderd%2Fnotifications%2Fenqueuer.go%60%5Cn++++%5B1d%5D%29%5Cn%29%22%2C%22queryType%22%3A%22range%22%2C%22datasource%22%3A%7B%22type%22%3A%22loki%22%2C%22uid%22%3A%22loki%22%7D%2C%22editorMode%22%3A%22code%22%2C%22step%22%3A%221d%22%7D%5D%2C%22range%22%3A%7B%22from%22%3A%221728803290436%22%2C%22to%22%3A%221731398890436%22%7D%2C%22panelsState%22%3A%7B%22logs%22%3A%7B%22visualisationType%22%3A%22logs%22%7D%7D%7D%7D&orgId=1)

The prospect of creating a completely new test implementation backed by a real `database.Store` was more than I could stomach right now, so I wired in some minimal RBAC assertions and fixed what tests ended up breaking.

Hoever, this is not the ideal fix. Changes to our RBAC policy on notifications will not be captured here and may result in similar sadness in future.

### CODE REVIEW

File: coderd/notifications/notificationstest/fake_enqueuer.go
Comment (johnstcn) [inline coderd/notifications/notificationstest/fake_enqueuer.go:1]: review: moved to its own package to avoid import cycle

File: coderd/autobuild/lifecycle_executor_test.go
Comment (johnstcn) [inline coderd/autobuild/lifecycle_executor_test.go:1]: review: no changes are needed to `lifecycle_executor.go` as it has its own actor to which I've added the required permissions

File: testutil/notifications.go
Comment (johnstcn) [inline testutil/notifications.go:1]: review: moved to `notificationstest`

**Review by johnstcn — COMMENTED** [review]
self-review: Another option I'm considering is just moving the dbauthz.AsNotifier calls inside the notifications enqueuer. That way, external callers don't need to think about authz when they enqueue a notification. This is both a pro and a con IMO. If folks feel strongly enough about it being the other way, I can make the necessary modifications.

File: coderd/workspaces.go
Comment (mtojek) [inline coderd/workspaces.go:1058]: Is there an option to place `dbauthz.AsNotifier` inside `Enqueue()` or add some wrapper on `NotificationsEnqueuer`?

File: coderd/notifications/notificationstest/fake_enqueuer.go
Comment (mtojek) [inline coderd/notifications/notificationstest/fake_enqueuer.go:58]: nit: should we call `Enqueue` instead?

File: coderd/notifications/notificationstest/fake_enqueuer.go
Comment (mtojek) [inline coderd/notifications/notificationstest/fake_enqueuer.go:99]: nit: are you creating a new slice to prevent modifications/dev errors?

**Review by mtojek — APPROVED** [review]

File: coderd/autobuild/lifecycle_executor_test.go
Comment (dannykopping) [inline coderd/autobuild/lifecycle_executor_test.go:1070]: Why're we skipping this?

File: coderd/autobuild/lifecycle_executor_test.go
Comment (dannykopping) [inline coderd/autobuild/lifecycle_executor_test.go:1114]: Left-over from testing?

**Review by dannykopping — CHANGES_REQUESTED** [review]
Blocking merge since you have an approval already; I think leaving these in was not intentional?

File: coderd/autobuild/lifecycle_executor_test.go
Comment (johnstcn) [reply coderd/autobuild/lifecycle_executor_test.go:1070]: :facepalm:

File: coderd/notifications/notificationstest/fake_enqueuer.go
Comment (johnstcn) [reply coderd/notifications/notificationstest/fake_enqueuer.go:99]: Yep, correct.

File: coderd/workspaces.go
Comment (johnstcn) [reply coderd/workspaces.go:1058]: Yes, that was an alternative option I was considering instead of forcing all of this onto callers. We can just have the notifier use the correct authz context internally and not have to worry about it. However, this means that _any_ actor can then enqueue _any_ kind of notification as long as it's plumbed through in code.

File: coderd/autobuild/lifecycle_executor_test.go
Comment (mtojek) [reply coderd/autobuild/lifecycle_executor_test.go:1070]: sorry, missed it

File: coderd/workspaces.go
Comment (johnstcn) [reply coderd/workspaces.go:1058]: I'll punt the decision for now and make it a potential follow-up since I'm not sure of the wider ramifications of this potential change yet.

Comment (DanielleMaywood) [reply coderd/workspaces.go:1058]: I think if we put `dbauthz.AsNotifier` inside `Enqueue` we're effectively working against the purpose of the RBAC system in my opinion.

File: coderd/notifications/notificationstest/fake_enqueuer.go
Comment (johnstcn) [reply coderd/notifications/notificationstest/fake_enqueuer.go:58]: The original implementation called `EnqueueWithData`:

```
func (f *FakeNotificationsEnqueuer) Enqueue(ctx context.Context, userID, templateID uuid.UUID, labels map[string]string, createdBy string, targets ...uuid.UUID) (*uuid.UUID, error) {
	return f.EnqueueWithData(ctx, userID, templateID, labels, nil, createdBy, targets...)
}
```

File: coderd/workspaces.go
Comment (mtojek) [reply coderd/workspaces.go:1058]: If only we can build a _magic_ wrapper that can assume the auth rule...

File: coderd/notifications/notificationstest/fake_enqueuer.go
Comment (mtojek) [reply coderd/notifications/notificationstest/fake_enqueuer.go:58]: sorry `EnqueueWithData` of course, but question stays the same

Comment (johnstcn) [reply coderd/notifications/notificationstest/fake_enqueuer.go:58]: Ah I see what you mean. Right now we do `Enqueue` -> `enqueueWithDataNoLock`. You're suggesting we do `Enqueue` -> `EnqueueWithData` -> lock -> `enqueueWithDataLock` or similar?

**Comment by dannykopping** [comment]
> self-review: Another option I'm considering is just moving the dbauthz.AsNotifier calls inside the notifications enqueuer. That way, external callers don't need to think about authz when they enqueue a notification. This is both a pro and a con IMO. If folks feel strongly enough about it being the other way, I can make the necessary modifications.

Why do you think it'd be a con? I think this would make sense and express the intent clearly.

File: coderd/notifications/notificationstest/fake_enqueuer.go
Comment (johnstcn) [reply coderd/notifications/notificationstest/fake_enqueuer.go:79]: Nice catch! This is an issue with the original implementation:

```
func (f *FakeNotificationsEnqueuer) EnqueueWithData(_ context.Context, userID, templateID uuid.UUID, labels map[string]string, data map[string]any, createdBy string, targets ...uuid.UUID) (*uuid.UUID, error) {
	f.mu.Lock()
	defer f.mu.Unlock()

	f.Sent = append(f.Sent, &Notification{
		UserID:     userID,
		TemplateID: templateID,
		Labels:     labels,
		Data:       data,
		CreatedBy:  createdBy,
		Targets:    targets,
	})

	id := uuid.New()
	return &id, nil
}
```

File: coderd/notifications/notificationstest/fake_enqueuer.go
Comment (johnstcn) [reply coderd/notifications/notificationstest/fake_enqueuer.go:29]: Would you be happy to leave this as a follow-up?

**Comment by johnstcn** [comment]
> > self-review: Another option I'm considering is just moving the dbauthz.AsNotifier calls inside the notifications enqueuer. That way, external callers don't need to think about authz when they enqueue a notification. This is both a pro and a con IMO. If folks feel strongly enough about it being the other way, I can make the necessary modifications.
> 
> Why do you think it'd be a con? I think this would make sense and express the intent clearly.

I could imagine later-on creating an endpoint `/api/v2/notifications/send` that does something similar to the below and does not perform any additional RBAC checks:

```
func (a *api) postNotification(rw http.ResponseWriter, r *http.Request) {
  [...]
  var req codersdk.NotificationRequest
  if err := httpapi.Read(&ctx, rw, r, &req); err != nil { ... }
  // Note: passing r.Context() here so that RBAC is enforced based on actor in context
  notifID, err := api.NotificationsEnqueuer.Enqueue(r.Context(), req.TemplateID, req.Labels, req.CreatedBy, req.Targets...)
  if err != nil { ... }
  httpapi.Write(r.Context(), http.StatusCreated, codersdk.Response{...}}
}
```

If `dbauthz.AsNotifier` is simply called within the notifier itself then all users can notify all users, and effectively no actor-level RBAC checking is done. If it's done at the API endpoint level then only users with the correct privileges are able to do so.

File: coderd/notifications/notificationstest/fake_enqueuer.go
Comment (dannykopping) [reply coderd/notifications/notificationstest/fake_enqueuer.go:29]: I would, I just hope we don't leave it too long in this state.

Comment (johnstcn) [reply coderd/notifications/notificationstest/fake_enqueuer.go:29]: Filed https://github.com/coder/coder/issues/15481 to follow up.

File: coderd/notifications/notificationstest/fake_enqueuer.go
Comment (johnstcn) [reply coderd/notifications/notificationstest/fake_enqueuer.go]: Done as suggested

**Review by mtojek — APPROVED** [review]
👍 

Before you all start going deeper into discussions keep in mind that firstly we need to mitigate the issue.

File: coderd/notifications/notificationstest/fake_enqueuer.go
Comment (dannykopping) [reply coderd/notifications/notificationstest/fake_enqueuer.go:29]: Thank you!!

File: coderd/notifications/notificationstest/fake_enqueuer.go
Comment (dannykopping) [reply coderd/notifications/notificationstest/fake_enqueuer.go:79]: Oh weird... indeed it is.

**Review by dannykopping — APPROVED** [review]
TYVM for addressing this! 🦸‍♂️

**Comment by johnstcn** [comment]
Local smoke-testing:

- Run `nc -kdl 8888` in a separate terminal
- Run `./scripts/develop.sh -- --notifications-method=webhook --notifications-webhook-endpoint=http://localhost:8888 --notifications-max-send-attempts=1`
- Users:
  - Create a user, suspend a user, resume a user
  - Observe notification message webhook body in netcat output
- Workspaces:
  - Create a workspace, enable dormancy on template
  - In local database: `UPDATE workspaces SET last_used_at = NOW() - INTERVAL '1 hours';`
  - Wait for dormancy to kick in
  - Observe notification webhook message body in netcat output

**Pull request merged**
